src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp

```diff
@@ -74,13 +77,527 @@ namespace
     const Class_ID AppleseedRendererClassId(0x72651b24, 0x5da32e1d);
     const Class_ID AppleseedRendererTabClassId(0x6155c0c, 0xed6475b);
     const wchar_t* AppleseedRendererClassName = L"appleseed Renderer";
+
+    enum ParamMapId
+    {
+        ParamMapIdOutput,
+        ParamMapIdImageSampling,
+        ParamMapIdLighting,
+        ParamMapIdSystem
+    };
+
+    enum ParamId
+    {
+        ParamIdOuputMode                = 0,
+        ParamIdProjectPath              = 1,
+        ParamIdScaleMultiplier          = 2,
+        
+        ParamIdPixelSamples             = 3,
+        ParamIdTileSize                 = 4,
+        ParamIdPasses                   = 5,
+        ParamIdFilterSize               = 6,
+        ParamIdFilterType               = 7,
+        
+        ParamIdEnableGI                 = 8,
+        ParamIdGIBounces                = 9,
+        ParamIdEnagleCaustics           = 10,
+        ParamIdEnableMaxRay             = 11,
+        ParamIdMaxRayValue              = 12,
+        ParamIdForceDefaultLightsOff    = 13,
+        ParamIdEnableBackgroundLight    = 14,
+        ParamIdBackgroundAlphaValue     = 15,
+
+        ParamIdCPUCores                 = 16,
+        ParamIdOpenLogMode              = 17,
+        ParamIdLogMaterialRendering     = 18,
+        ParamIdUseMaxProcedurals        = 19,
+        ParamIdEnableLowPriority        = 20,
+        ParamIdEnableRenderStamp        = 21,
+        ParamIdRenderStampFormat        = 22,
+    };
+
+    MCHAR* GetString(int id)
+    {
+        static wchar_t buf[256];
```

Comment:
> I thought static makes it safe, no?

I'm not sure what you mean by "safe". `static` just means that a single buffer is allocated, instead of one per invocations of `GetString()`. Declaring `buf` as static is the same as moving its declaration outside of `GetString()`. What this means is that in the following code:

    const MCHAR* s0 = GetString(0);
    const MCHAR* s1 = GetString(1);

`s0` and `s1` are equal, they both point to `buf`; as such the second call to `GetString()` invalidates the value returned by the first call.